src/aggregation/bucket/histogram/histogram.rs

```diff
@@ -0,0 +1,822 @@
+/// DistributedAggregationCollector.
+///
+/// # Request JSON Format
+/// ```ignore
+/// {
+///     "prices": {
+///         "histogram": {
+///             "field": "price",
+///             "interval": 10,
+///         }
+///     }
+/// }
+///  ```
+#[derive(Clone, Debug, Default, PartialEq, Serialize, Deserialize)]
+pub struct HistogramAggregation {
+    /// The field to aggregate on.
+    pub field: String,
+    /// The interval to chunk your data range. The buckets span ranges of [0..interval).
+    /// Must be a positive value.
+    pub interval: f64,
+    /// Intervals intersect at 0 by default, offset can move the interval.
+    /// Offset has to be in the range [0, interval).
+    ///
+    /// As an example. If there are two documents with value 8 and 12 and interval 10.0, they would
+    /// fall into the buckets with the key 0 and 10.
+    /// With offset 5 and interval 10, they would both fall into the bucket with they key 5 and the
+    /// range [5..15)
+    pub offset: Option<f64>,
+    /// The minimum number of documents in a bucket to be returned. Defaults to 0.
+    pub min_doc_count: Option<u64>,
+    /// Sets a hard limit for the data range.
+    /// This can be used to filter values if they are not in the data range.
+    pub hard_bounds: Option<HistogramBounds>,
+    /// Can be set to extend your bounds. The range of the buckets is by default defined by the
+    /// data range of the values of the documents. As the name suggests, this can only be used to
+    /// extend the value range. If the bounds for min or max are not extending the range, the value
+    /// has no effect on the returned buckets.
+    ///
+    /// Cannot be set in conjunction with min_doc_count > 0, since the empty buckets from extended
+    /// bounds would not be returned.
+    pub extended_bounds: Option<HistogramBounds>,
+}
+
+impl HistogramAggregation {
+    fn validate(&self) -> crate::Result<()> {
+        if self.interval <= 0.0f64 {
+            return Err(TantivyError::InvalidArgument(
+                "interval must be a positive value".to_string(),
+            ));
+        }
+
+        if self.min_doc_count.unwrap_or(0) > 0 && self.extended_bounds.is_some() {
+            return Err(TantivyError::InvalidArgument(
+                "Cannot set min_doc_count and extended_bounds at the same time".to_string(),
+            ));
+        }
+
+        Ok(())
+    }
+
+    /// Returns the minimum number of documents required for a bucket to be returned.
+    pub fn min_doc_count(&self) -> u64 {
+        self.min_doc_count.unwrap_or(0)
+    }
+}
+
+/// Used to set extended or hard bounds on the histogram.
+#[derive(Clone, Copy, Debug, PartialEq, Serialize, Deserialize)]
+pub struct HistogramBounds {
+    /// The lower bounds.
+    pub min: f64,
+    /// The upper bounds.
+    pub max: f64,
+}
+
+impl HistogramBounds {
+    fn contains(&self, val: f64) -> bool {
+        val >= self.min && val <= self.max
```

Comment:
What?

Reply:
oh yeah I think you are right, I have to check elastic on this